profiler/src/ProfilerEngine/Datadog.Profiler.Native.Windows/EtwEventsManager.cpp

```diff
@@ -0,0 +1,392 @@
+
+    // convert into system time
+    SYSTEMTIME st;
+    ::FileTimeToSystemTime(&ft, &st);
+
+    // take GMT shift into account
+    ::SystemTimeToTzSpecificLocalTime(nullptr, &st, &st);
+
+    // convert in epoch time
+    tm t = {0};
+    t.tm_year = st.wYear - 1900;
+    t.tm_mon = st.wMonth - 1;
+    t.tm_mday = st.wDay;
+    t.tm_hour = st.wHour;
+    t.tm_min = st.wMinute;
+    t.tm_sec = st.wSecond;
+    time_t timeSinceEpoch = mktime(&t);
+
+    return timeSinceEpoch * 1000'000'000;
+}
+
+void EtwEventsManager::OnEvent(
+    uint64_t systemTimestamp,
+    uint32_t tid,
+    uint32_t version,
+    uint64_t keyword,
+    uint8_t level,
+    uint32_t id,
+    uint32_t cbEventData,
+    const uint8_t* pEventData)
+{
+    if (keyword == KEYWORD_STACKWALK)
+    {
+        if (_isDebugLogEnabled)
+        {
+            //std::cout << "StackWalk" << std::endl;
+        }
+
+        auto pThreadInfo = GetOrCreate(tid);
+        if (pThreadInfo->LastEventWasContentionStart())
+        {
+            AttachContentionCallstack(pThreadInfo, cbEventData, pEventData);
+        }
+        else if (pThreadInfo->LastEventWasAllocationTick())
+        {
+            AttachAllocationCallstack(pThreadInfo, cbEventData, pEventData);
+        }
+
+        pThreadInfo->ClearLastEventId();
+    }
+    else if (keyword == KEYWORD_CONTENTION)
+    {
+        if (_pContentionListener == nullptr)
+        {
+            return;
+        }
+
+        if (id == EVENT_CONTENTION_START)
+        {
+            if (_isDebugLogEnabled)
+            {
+                std::cout << "ContentionStart" << std::endl;
+            }
+
+            auto pThreadInfo = GetOrCreate(tid);
+            pThreadInfo->ContentionStartTimestamp = TimestampToEpochNS(systemTimestamp);
+            pThreadInfo->LastEventId = EventId::ContentionStart;
+        }
+        else if (id == EVENT_CONTENTION_STOP)
+        {
+            if (_isDebugLogEnabled)
+            {
+                std::cout << "ContentionStop" << std::endl;
+            }
+
+            auto pThreadInfo = Find(tid);
+            if (pThreadInfo != nullptr)
+            {
+                pThreadInfo->ClearLastEventId();
+                if (pThreadInfo->ContentionStartTimestamp != 0)
+                {
+                    auto timestamp = TimestampToEpochNS(systemTimestamp);
+                    auto duration = timestamp - pThreadInfo->ContentionStartTimestamp;
+                    pThreadInfo->ContentionStartTimestamp = 0;
+
+                    _pContentionListener->OnContention(timestamp, tid, duration, pThreadInfo->ContentionCallStack);
+                }
+            }
+            else
+            {
+                // this should never happen
+            }
+        }
+    }
+    else if (keyword == KEYWORD_GC)
+    {
+
+        if (id == EVENT_ALLOCATION_TICK)
+        {
+            if (_isDebugLogEnabled)
+            {
+                //std::cout << "AllocationTick" << std::endl;
+            }
```

Comment:
can be removed if it's not needed

Reply:
I will clean up the verbose logs when everything works as expected